src/retry-interceptor.js

```diff
@@ -0,0 +1,111 @@
+
+  constructor(retryConfig?: RetryConfiguration) {
+    this.retryConfig = Object.assign({}, defaultRetryConfig, retryConfig || {});
+
+    if (this.retryConfig.strategy === retryStrategy.exponential &&
+       this.retryConfig.interval <= 1000) {
+      throw new Error('An interval less than or equal to 1 second is not allowed when using the exponential retry strategy');
+    }
+  }
+
+  request(request) {
+    if (!request.retryConfig) {
+      request.retryConfig = Object.assign({}, this.retryConfig);
+      request.retryConfig.counter = 0;
+    }
+    return request;
+  }
+
+  response(response, request) {
+    // retry was successful, so clean up after ourselves
+    delete request.retryConfig;
+    return response;
+  }
+
+  responseError(error, request) {
+    const { retryConfig } = request;
+    const { requestClone } = retryConfig;
+    return Promise.resolve().then(() => {
+      if (retryConfig.counter < retryConfig.maxRetries) {
+        const result = retryConfig.doRetry ? retryConfig.doRetry(error, request) : true;
+
+        return Promise.resolve(result).then(doRetry => {
+          if (doRetry) {
+            retryConfig.counter++;
+
+            return new Promise(resolve => setTimeout(resolve, calculateDelay(retryConfig) || 0)).then(() => {
+              let newRequest = requestClone.clone();
+              if (typeof (retryConfig.beforeRetry) === 'function') {
+                return retryConfig.beforeRetry(newRequest, retryConfig.client);
+              }
+              return newRequest;
+            }).then(newRequest => {
+              return retryConfig.client.fetch(Object.assign(newRequest, { retryConfig }));
+            });
+          }
+
+          // no more retries, so clean up
+          delete request.retryConfig;
+          throw error;
+        });
+      }
+      // no more retries, so clean up
+      delete request.retryConfig;
+      throw error;
+    });
+  }
+}
+
+function calculateDelay(retryConfig: RetryConfiguration) {
+  const { interval, strategy, minRandomInterval, maxRandomInterval, counter } = retryConfig;
+
+  if (typeof (strategy) === 'function') {
+    return retryConfig.strategy(counter);
+  }
+
+  switch (strategy) {
+  case (retryStrategy.fixed):
+    return retryStrategies[retryStrategy.fixed](interval);
+  case (retryStrategy.incremental):
+    return retryStrategies[retryStrategy.incremental](counter, interval);
+  case (retryStrategy.exponential):
+    return retryStrategies[retryStrategy.exponential](counter, interval);
+  case (retryStrategy.random):
+    return retryStrategies[retryStrategy.random](counter, interval, minRandomInterval, maxRandomInterval);
+  default:
+    throw new Error('Unrecognized retry strategy');
+  }
+}
+
+const retryStrategies = [
+  // fixed
+  interval => interval,
+
+  // incremental
+  (retryCount, interval) => interval * retryCount,
+
+  // exponential
+  (retryCount, interval) => retryCount === 1 ? interval : Math.pow(interval, retryCount) / 1000,
```

Comment:
@AshleyGrant Why divide out the thousand?

Reply:
The interval is in milliseconds. For example, 3 second interval: 3000^2 = 9,000,000 vs 3^2 = 9. I have to strip off 1,000 to the interval doesn't jump from 3 seconds to 2.5 hours.

Reply:
Yes but you'll want to exponentiate the 1000 as well.

3000 ^3 = 27 * (1000)^3.